### PR TITLE
Peer filter

### PR DESCRIPTION
**Pull request by ebuchman**
This PR introduces peer filters, enabling the app to determine which peers we should connect to based on their address and/or public key.

It depends on https://github.com/tendermint/go-p2p/pull/6

Changes:
- typed proxyAppConns
- new "query" conn - used for Query and Info
- call switch.SetAddrFilter and switch.SetPubKeyFilter

Issues:
- query conn: https://github.com/tendermint/tendermint/issues/254
- peer filtering: https://github.com/tendermint/tendermint/issues/244

Note:
Uses the first few commits from https://github.com/tendermint/tendermint/pull/265, but should be merged before that

### CODE REVIEW

**Comment by ebuchman** [comment]
TODO: turn this off by default from config so its not breaking

File: consensus/state.go [hunk 1]
```diff
@@ -1283,7 +1283,7 @@ func (cs *ConsensusState) commitStateUpdateMempool(s *sm.State, block *types.Blo
 	defer cs.mempool.Unlock()
 
 	// flush out any CheckTx that have already started
-	cs.proxyAppConn.FlushSync()
+	// cs.proxyAppConn.FlushSync() // ?! XXX
```
Comment (jaekwon) [inline consensus/state.go]: Yeah, this is handled by mempool/mempool.go first few lines in Update().  This should be removed.

File: proxy/state.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package proxy
```
Comment (jaekwon) [inline proxy/state.go]: Is this supposed to be here?

File: node/node.go [hunk 3]
```diff
@@ -118,13 +112,34 @@ func NewNode(config cfg.Config, privValidator *types.PrivValidator, getProxyApp
 	sw.AddReactor("BLOCKCHAIN", bcReactor)
 	sw.AddReactor("CONSENSUS", consensusReactor)
 
+	// filter peers by addr or pubkey with a tmsp query.
+	// if the query return code is OK, add peer
+	// XXX: query format subject to change
+	if config.GetBool("filter_peers") {
```
Comment (jaekwon) [inline node/node.go]: that's cool

File: proxy/state.go (hunk 2 above)
Comment (ebuchman) [inline proxy/state.go]: yeh. Makes more sense in the handshake PR which some of these commits were pulled from. empty interfaces here but filled in there

**Comment by jaekwon** [comment]
LGTM

**Pull request merged**